plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/VaultSecretSource.java

```diff
@@ -39,114 +37,200 @@
+        if (timeDiffInMillis < -5000L) {
+            // token will be valid for at least another 5s
+            result = false;
+            LOGGER.log(Level.FINE, "Auth token is still valid");
+        } else {
+            LOGGER.log(Level.FINE, "Auth token has to be re-issued" + timeDiffInMillis);
         }
 
-        // attempt AppRole login
-        if (vaultAppRole.isPresent() && vaultAppRoleSecret.isPresent() && !authToken.isPresent()) {
-            try {
-                authToken = Optional.ofNullable(
-                        vault.auth().loginByAppRole(vaultAppRole.get(), vaultAppRoleSecret.get()).getAuthClientToken()
-                );
-                LOGGER.log(Level.FINE, "Login to Vault using AppRole/SecretID successful");
-            } catch (VaultException e) {
-                LOGGER.log(Level.WARNING, "Could not login with AppRole", e);
-            }
+        return result;
+    }
+
+    private void setAuthTokenInVaultClient(String authToken) {
+        // Set auth token
+        try {
+            vaultConfig.token(authToken).build();
+        } catch (VaultException e) {
+            LOGGER.log(Level.WARNING, "Could not set auth token in vault client", e);
         }
 
-        // attempt User/Pass login
-        if (vaultUser.isPresent() && vaultPw.isPresent() && !authToken.isPresent()) {
-            try {
-                authToken = Optional.ofNullable(
-                        vault.auth().loginByUserPass(vaultUser.get(), vaultPw.get(), vaultMount.get()).getAuthClientToken()
-                );
-                LOGGER.log(Level.FINE, "Login to Vault using User/Pass successful");
-            } catch (VaultException e) {
-                LOGGER.log(Level.WARNING, "Could not login with User/Pass", e);
-            }
+        // Set token expiration date
+        try {
+            authTokenExpiration = Calendar.getInstance();
+            authTokenExpiration.add(Calendar.SECOND, (int) vault.auth().lookupSelf().getTTL());
+        } catch (VaultException e) {
+            LOGGER.log(Level.WARNING, "Could not determine token expiration. Assuming expired soon.", e);
+            authTokenExpiration = Calendar.getInstance();
         }
+    }
 
-        // Use authToken to read secrets from vault
-        if (authToken.isPresent()) {
-            readSecretsFromVault(authToken.get(), vaultConfig, vault, vaultPaths.get());
-        } else {
-            LOGGER.log(Level.WARNING, "Vault auth token missing. Cannot read from vault");
+    private void authenticate() {
+        Optional<String> vaultTokenOpt = Optional.ofNullable(vaultToken);
+        Optional<String> vaultAppRoleOpt = Optional.ofNullable(vaultAppRole);
+        Optional<String> vaultAppRoleSecretOpt = Optional.ofNullable(vaultAppRoleSecret);
+        Optional<String> vaultUserOpt = Optional.ofNullable(vaultUser);
+        Optional<String> vaultPwOpt = Optional.ofNullable(vaultPw);
+
+        if (!(vaultTokenOpt.isPresent() ||
```

Comment:
can you break this condition down into either multiple if statements, extracted methods or something

Reply:
Restructured logic to place authentication logic outside